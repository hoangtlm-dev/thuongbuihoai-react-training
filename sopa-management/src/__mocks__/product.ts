const MOCK_PRODUCTS = [
  {
    name: 'Black shoes',
    price: 87,
    size: ['Small', 'Medium', 'Large'],
    description:
      'Our goal in designing the super-soft Atoms Crew Socks was to provide a seamless match with your Atoms shoes, enhancing your comfort for everyday wear - making them the perfect socks for your everyday shoes.',
    color: 'Black',
    image: {
      url: '/images/products/black-shoes.png',
      alt: 'This is black shoes'
    },
    id: '1'
  },
  {
    name: 'Blue shoes',
    price: 45,
    size: ['Small', 'Medium', 'Large'],
    description:
      'Our goal in designing the super-soft Atoms Crew Socks was to provide a seamless match with your Atoms shoes, enhancing your comfort for everyday wear - making them the perfect socks for your everyday shoes.',
    color: 'Blue',
    image: {
      url: '/images/products/blue-shoes.png',
      alt: 'This is blue shoes'
    },
    id: '2'
  }
];

const MOCK_PRODUCT = {
  name: 'Black shoes',
  price: 87,
  size: ['Small', 'Medium', 'Large'],
  description:
    'Our goal in designing the super-soft Atoms Crew Socks was to provide a seamless match with your Atoms shoes, enhancing your comfort for everyday wear - making them the perfect socks for your everyday shoes.',
  color: 'Black',
  image: {
    url: '/images/products/black-shoes.png',
    alt: 'This is black shoes'
  },
  id: '1'
};

export {
  MOCK_PRODUCTS,
  MOCK_PRODUCT
};
