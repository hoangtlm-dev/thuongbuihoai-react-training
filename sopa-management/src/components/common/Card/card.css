.card {
  width: var(--w-medium);
  height: var(--h-large);
  border-radius: var(--rounded-primary);
  background: var(--bg-primary);
  margin-bottom: 30px;
}

.card-image {
  text-align: center;
  margin-top: 80px;
}

.card-item {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  height: var(--h-medium);
  padding: var(--px-medium) var(--py-normal) var(--py-normal) var(--py-normal);
}

.card-description {
  display: flex;
  margin-top: 8px;
  justify-content: space-between;
}

.card-name {
  color: var(--dark);
  font-size: var(--text-extra-sm);
  line-height: var(--leading-primary);
  font-weight: var(--font-medium);
  text-decoration: none;
}
