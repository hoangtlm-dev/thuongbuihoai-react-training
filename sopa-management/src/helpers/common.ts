// Flattens an array of arrays into a single array
export const flattenArray = <T>(pages: T[][]): T[] => {
  const result: T[] = [];
  pages.forEach((page) => {
    result.push(...page);
  });

  return result;
};
