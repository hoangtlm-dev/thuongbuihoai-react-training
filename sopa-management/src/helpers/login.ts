import {
  Account,
  LoginType
} from "@interfaces/account";
import { FieldError } from "react-hook-form";

export enum CheckType {
  login = 'login',
  email = 'email',
  password = 'password'
}

// Check login
export const checkAccount = (
  data?: Account[],
  email?: string,
  password?: string,
  checkType?: CheckType,
) => {
  switch (checkType) {
    case CheckType.login:
      return data?.some((user: Account) => user.email === email && user.password === password);
    case CheckType.email:
      return data?.some((user: Account) => user.email === email);
    case CheckType.password:
      return data?.some((user: Account) => user.password === password);
    default:
      throw new Error('Invalid check type');
  }
};

export const checkValidationStyles = (
  serverError: boolean,
  rules: FieldError | undefined,
  isDirty: boolean,
  type: LoginType,
) => {
  const checkIsDirty = isDirty ? type.info : type.default
  const style = serverError || rules ? type.error : checkIsDirty;

  return { style };
};