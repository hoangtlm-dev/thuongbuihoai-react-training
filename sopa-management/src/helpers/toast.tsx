import Toast from "../components/Toast";

export const showToast = (
  title: string,
  message: string,
  type: 'success' | 'error',
  duration: number
) => {
  return (
    <Toast
      title={title}
      message={message}
      type={type}
      duration={duration}
    />
  )
};
