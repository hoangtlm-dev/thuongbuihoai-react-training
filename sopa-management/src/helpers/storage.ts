import { Product } from "../interfaces/product";

export const setStorage = (key: string, data: Product) => {
  localStorage.setItem(key, JSON.stringify(data));
};

export const getStorage = (key: string) => {
  return JSON.parse(localStorage.getItem(key) as any) || [];
};

export const clearStorage = () => {
  localStorage.clear();
};
