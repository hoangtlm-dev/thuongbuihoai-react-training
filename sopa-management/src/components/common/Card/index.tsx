import { memo } from 'react';
import { Link } from 'react-router-dom';

// Components
import Text,
{
  ThemeType
} from '@common/Text';
import Price,
{
  PriceType
} from '@common/Price';
import Image from '@common/Image';

// Styles
import './card.css';

interface CardProps {
  key?: string;
  name?: string;
  color?: string;
  price?: number;
  href: string;
  src: string;
  alt?: string;
}

const Card: React.FC<CardProps> = ({
  name = '',
  color = '',
  price = 0,
  href,
  src,
  alt = ''
}) => {
  return (
    <div data-testid='card' className='card'>
      <div className='card-item'>
        <figure className='card-image'>
          <Image href={href} url={src} alt={alt} className='card-image-item' />
        </figure>
        <div className='card-info'>
          <Link to={href} className='card-name'>
            {name}
          </Link>
          <div className='card-description'>
            <Text text={color} type={ThemeType.highlightPrimary} />
            <Price value={price} type={PriceType.primary} />
          </div>
        </div>
      </div>
    </div>
  );
};

export default memo(Card);
