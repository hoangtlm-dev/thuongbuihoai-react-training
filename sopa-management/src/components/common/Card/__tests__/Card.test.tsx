// Libs
import '@testing-library/jest-dom';

// Mocks
import { MOCK_PRODUCT } from '../../../../__mocks__/product';

// Components
import { renderRouterTest } from '../../../../helpers/testUtils';
import Card from '..';

describe('Card component', () => {
  test('should render Card component', () => {
    const { getByTestId } = renderRouterTest(
      <Card
        href='/'
        src={MOCK_PRODUCT.image?.url}
        name={MOCK_PRODUCT.name}
        color={MOCK_PRODUCT.color}
        price={MOCK_PRODUCT.price}
      />
    );
    expect(getByTestId('card')).toBeInTheDocument();
  });

  test('should render Card component by default', () => {
    const { getByTestId } = renderRouterTest(
      <Card href='/' src={MOCK_PRODUCT.image?.url} />
    );
    expect(getByTestId('card')).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const { asFragment } = renderRouterTest(
      <Card
        href='/'
        src={MOCK_PRODUCT.image?.url}
        name={MOCK_PRODUCT.name}
        color={MOCK_PRODUCT.color}
        price={MOCK_PRODUCT.price}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
