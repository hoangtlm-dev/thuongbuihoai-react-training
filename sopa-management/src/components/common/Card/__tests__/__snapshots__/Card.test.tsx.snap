// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="card"
    data-testid="card"
  >
    <div
      class="card-item"
    >
      <figure
        class="card-image"
      >
        <a
          data-testid="image-link"
          href="/"
        >
          <img
            alt=""
            class="card-image-item"
            src="/images/products/black-shoes.png"
          />
        </a>
      </figure>
      <div
        class="card-info"
      >
        <a
          class="card-name"
          href="/"
        >
          Black shoes
        </a>
        <div
          class="card-description"
        >
          <p
            class=" text-highlight-primary"
            data-testid="text"
          >
            Black
          </p>
          <p
            class=" price-primary"
            data-testid="price"
          >
            $87
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
